crates/kubelet/src/volume/persistentvolumeclaim.rs

```diff
@@ -0,0 +1,319 @@
+use std::error::Error;
+use std::fmt;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use k8s_csi::v1_3_0::node_client::NodeClient;
+use k8s_csi::v1_3_0::node_service_capability::{rpc, Rpc, Type as CapabilityType};
+use k8s_csi::v1_3_0::volume_capability::access_mode::Mode as CSIMode;
+use k8s_csi::v1_3_0::volume_capability::{
+    AccessMode as CSIAccessMode, AccessType as CSIAccessType, MountVolume as CSIMountVolume,
+};
+use k8s_csi::v1_3_0::{
+    NodeGetCapabilitiesRequest, NodePublishVolumeRequest, NodeStageVolumeRequest, VolumeCapability,
+};
+
+use k8s_openapi::api::core::v1::{PersistentVolume, PersistentVolumeClaimSpec};
+use k8s_openapi::api::storage::v1::StorageClass;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::LabelSelector;
+
+use crate::grpc_sock;
+use crate::plugin_watcher::PluginRegistry;
+
+use super::*;
+
+// https://github.com/kubernetes/kubernetes/blob/0b1d94a849e3e31d83b8e29454bf6698b93a8f11/pkg/volume/csi/csi_plugin.go#L54
+const VOLUME_DATA_FILE_NAME: &str = "vol_data.json";
+
+#[derive(Debug, Clone)]
+struct BadVolumeMode;
+
+impl fmt::Display for BadVolumeMode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "bad volume mode")
+    }
+}
+
+impl Error for BadVolumeMode {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        None
+    }
+}
+
+// Kubernetes supports two volumeModes of PersistentVolumes: Filesystem and
+// Block.
+//
+// volumeMode is an optional API parameter. Filesystem is the default mode used
+// when volumeMode parameter is omitted.
+//
+// A volume with volumeMode: Filesystem is mounted into Pods into a directory.
+// If the volume is backed by a block device and the device is empty,
+// Kuberneretes creates a filesystem on the device before mounting it for the
+// first time.
+//
+// You can set the value of volumeMode to Block to use a volume as a raw block
+// device. Such volume is presented into a Pod as a block device, without any
+// filesystem on it. This mode is useful to provide a Pod the fastest possible
+// way to access a volume, without any filesystem layer between the Pod and the
+// volume. On the other hand, the application running in the Pod must know how
+// to handle a raw block device. See Raw Block Volume Support for an example on
+// how to use a volume with volumeMode: Block in a Pod.
+//
+// https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-mode
+#[derive(Debug)]
+enum VolumeMode {
+    Block,
+    Filesystem,
+}
+
+impl FromStr for VolumeMode {
+    type Err = BadVolumeMode;
+
+    // defines what type of volume is required by the claim. The "filesystem"
+    // mode is implied when not included in the claim spec.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "Block" => Ok(VolumeMode::Block),
+            "Filesystem" => Ok(VolumeMode::Filesystem),
+            "" => Ok(VolumeMode::Filesystem),
+            _ => Err(BadVolumeMode),
```

Comment:
Is it possible for these to be created with the other values? Is there value in creating these new types rather than just doing `anyhow::bail` with some context?